core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -2877,9 +2877,7 @@ private void optimiseOrderBy(QueryModel model, int topLevelOrderByMnemonic) {
         for (int i = 0, k = jm.size(); i < k; i++) {
             QueryModel qm = jm.getQuick(i).getNestedModel();
             if (qm != null) {
-                if (model.getGroupBy().size() == 0
-                        && model.getSampleBy() == null
-                        && model.getSelectModelType() != QueryModel.SELECT_MODEL_DISTINCT) { // order by should not copy through group by, sample by or distinct
```

Comment:
I think it was because propagating order by would end up pushing it down while it should remain on the same level . 
```sql
select distinct x from (tab) order by x 
```
vs 
```sql
select distinct x from (tab order by x)
```

Reply:
I see. I wonder if we have any tests around that and why they didn't break because of this change.

Reply:
Thanks for the feedback. I'm analyzing this :).

Reply:
My reasoning is as follows:
1. By pushing down ORDER BY advice, we potentially influence the result returned by `io.questdb.cairo.sql.RecordCursorFactory#followedOrderByAdvice`. Thus, a cursor factory that used to return false from this method before my change may now potentially return true. I assume that `followedOrderByAdvice` is the single source of truth telling whether order advice has been pushed down and used by a given factory.
2. There are the following factories that provide support for DISTINCT:
   * `DistinctSymbolRecordCursorFactory`
   * `DistinctIntKeyRecordCursorFactory`
   * `DistinctTimeSeriesRecordCursorFactory`
   * `DistinctRecordCursorFactory`
   None of them overwrites the default implementation of `followedOrderByAdvice`, which always returns false. 
3. Looking at the hierarchy in `io.questdb.griffin.SqlCodeGenerator#generateQuery0` and knowing that a distinct cursor factory is created at the bottom of it, I deduce that the ORDER BY clause is always applied to the DISTINCT model with which it is associated.
4. Now, looking at `io.questdb.griffin.SqlCodeGenerator#generateOrderBy`, I see that sorting is entirely skipped if at least one of the following is true:
   * The method `followedOrderByAdvice` of the base factory returned true. This is not the case here because DISTINCT factories always return false.
   * The ORDER BY clause is empty. This means that the ORDER BY advice is also empty.
   * The ORDER BY clause has only one element - the designated timestamp, and the base factory has the scan direction aligned with the order from the clause. My change affects this scenario - due to the advice, the scan direction is aligned with the order from the clause.

   In other cases, some kind of sorting is applied.

Do you see any flaws in my reasoning?

Reply:
That's correct reasoning. I don't see any problems in it. @bziobrowski WDYT?

To make sure that no regression arise in future, it might be a great idea to add a bunch of tests around SELECT DISTINCT + ORDER BY, hitting all existing distinct factories. We have some in `DistinctWithLimitTest`, but maybe some cases are missing.

Reply:
@piotrrzysko  It looks good to me . 

Re 3. - ORDER BY is ignored in many cases (e.g. 
```sql
select * from (select * from tab order by x) order by y
```
normally skips the `order by x`.
It'd be good to add tests of distinct over all join types.

Reply:
@piotrrzysko to make solving this problem easier I would suggest to break this problem into two parts:

1. Implementing `distinct/order by/limit`, e.g. distinct algo that is both order and limit aware in isolation. This could be a new factory.
2. Wiring this algorithm into the existing optimiser/generator.

Those are relatively self contained problems. If you start with 1 you will inevitably learn about 2. I would also discourage strarting with 2, which is what you're embarking on.

Does this make sense?

Reply:
OK Vlad, I'll try to follow your advice (no pun intended ;)).

Reply:
It might be simpler to tweak `DistinctTimeSeriesRecordCursorFactory`, so that it's capable of "stealing" LIMIT clause in case of no ORDER BY or ORDER BY ts ASC/DESC. We do something similar in `AsyncFilteredRecordCursorFactory` (this is flagged by the `followedLimitAdvice()` method).

Reply:
First of all, thanks again for the feedback. I tried to compile all the information you provided me, and here is what I got:

Starting from the bottom. In order to process as few rows as possible, it is necessary to scan a table in the direction prescribed by the ORDER BY clause. If I were to implement a new factory, I would need to either start at a very low level (presumably at the level of `DataFrameCursorFactory`) and implement features like filtering in the factory, or make the base record factory scan the table in the proper direction. The former approach seemed overkill, so I thought that the latter one would be better. I realized that at the level of record cursor factories, there are two options to change the direction of table scanning:

* The getPageFrameCursor method that takes the order as a parameter. However, not all record factories support framing. For example, `AsyncJitFilteredRecordCursorFactory` doesn't support it.
* Propagating ORDER BY advice.

Again, the first option didn't seem applicable to my case. Perhaps I could introduce a new mechanism for that, but I decided to give ORDER BY advice another chance.

Then, assuming that I could obtain records in the desired order, I realized adding support for ORDER BY and LIMIT to the new factory is not a big challenge because (as Andrei suggested) I could just extend `DistinctTimeSeriesRecordCursorFactory` to be aware of the LIMIT clause, and I'm done.

After that, I proceeded to plug my factory into the optimizer/generator. I realized that instead of making `DistinctTimeSeriesRecordCursorFactory` aware of the LIMIT clause, I could leverage existing factories (in particular `LimitRecordCursorFactory`). The only requirement I needed to meet was that `DistinctTimeSeriesRecordCursor` returns records in the desired order, which it does as long as the base factory is told to do so. Thereby, I ended up in the place where I started, trying to push down the ORDER BY advice. BUT, along the way, I learned that my original approach of affecting all DISTINCT queries wasn't safe, and I should have narrowed down the scope of the change. Hence, I tweaked the condition in the optimizer to be stricter and target only cases that are handled by `DistinctTimeSeriesRecordCursorFactory`.


Additionally:
* I've added several tests that verify pushing down the ORDER BY clause is not applied to cases that do not concern this task (I hope I applied all Andrei's and Bolek's suggestions + added a few additional test cases).
* As a byproduct, I optimized sorting for queries like: `SELECT DISTINCT ts, [col1, col2..] FROM tab ORDER BY ts` (no limit).

Looking forward to your opinion on my approach. Please, let me know if I'm missing something.

Reply:
This sounds great! Let me check the changes.

Reply:
Let me resolve this lengthy thread. Let's continue with the new comments.